src/ip_notifier.cpp

```diff
@@ -55,11 +58,105 @@ namespace
 	// TODO: ip_change_notifier_sim
 #elif TORRENT_USE_NETLINK
 	// TODO: ip_change_notifier_nl
+#elif TORRENT_USE_SYSTEMCONFIGURATION
+// see https://developer.apple.com/library/content/technotes/tn1145/_index.html
+
+CFArrayRef create_keys_array()
+{
+	CFMutableArrayRef keys = CFArrayCreateMutable(nullptr
+		, 0, &kCFTypeArrayCallBacks);
+	CFStringRef key;
+
+	// "State:/Network/Interface/[^/]+/IPv4"
+	key = SCDynamicStoreKeyCreateNetworkInterfaceEntity(nullptr
+		, kSCDynamicStoreDomainState, kSCCompAnyRegex, kSCEntNetIPv4);
+	CFArrayAppendValue(keys, key);
+	CFRelease(key);
+
+	// NOTE: for IPv6, you can replicate the above setup with kSCEntNetIPv6
+	// but due to the current state of most common configurations, where
+	// IPv4 is used alongside with IPv6, you will end up with twice the
+	// notifications for the same change
+
+	return keys;
+}
+
+SCDynamicStoreRef create_dynamic_store(SCDynamicStoreCallBack callback, void* context_info)
+{
+	TORRENT_ASSERT(callback != nullptr);
+
+	SCDynamicStoreContext context = {0, nullptr, nullptr, nullptr, nullptr};
+	context.info = context_info;
+	SCDynamicStoreRef store = SCDynamicStoreCreate(nullptr
+		, CFSTR("libtorrent.IPChangeNotifierStore"), callback, &context);
+	if (store == nullptr)
+		return nullptr;
+
+	CFArrayRef keys = create_keys_array();
+	bool success = SCDynamicStoreSetNotificationKeys(store, nullptr, keys);
+	CFRelease(keys);
+
+	return success ? store : nullptr;
+}
+
+struct ip_change_notifier_macos final : ip_change_notifier
+{
+	explicit ip_change_notifier_macos(io_service& ios)
+		: m_ios(ios)
+	{
+		m_queue = dispatch_queue_create("libtorrent.IPChangeNotifierQueue", nullptr);
```

Comment:
it's my understanding that ARC takes care of that, in fact, you can't call `CFRelease` on it. Do you mind elaborate the "owned by a handle" idea?